### PR TITLE
Load shared libraries locally

### PR DESCRIPTION
**Pull request by ryanking13**
### Description

Fixes issues with symbol resolution and changes the behavior of loading shared libraries globally to local.

I found several libraries misconfigured while working on this: not having correct dependencies, so this PR fixes them as well.

### Checklists

- [x] Add a [CHANGELOG](https://github.com/pyodide/pyodide/blob/main/docs/project/changelog.md) entry

### CODE REVIEW

File: emsdk/patches/0006-Make-RTLD_LOCAL-work-correctly-for-preloaded-DSOs-21.patch
Comment (hoodmane) [inline emsdk/patches/0006-Make-RTLD_LOCAL-work-correctly-for-preloaded-DSOs-21.patch:1]: So this patch will go into Emscripten 3.1.62.

File: packages/cryptography/meta.yaml [hunk 1]
```diff
@@ -10,15 +10,9 @@ source:
...
-    export OPENSSL_LIBRARY_PATH=$(pkg-config --libs-only-L --dont-define-prefix openssl)
     export OPENSSL_DIR=$WASM_LIBRARY_DIR
-  cflags: |
-    -Wno-implicit-function-declaration
-    $(OPENSSL_INCLUDE_PATH)
-  ldflags: |
-    $(OPENSSL_LIBRARY_PATH)
-    -Wl,--no-entry
+    # Allow linking to native libraries, otherwise libssl and libcrypto will be not linked to the final binary
+    export RUSTFLAGS="-Z link-native-libraries=yes"
```
Comment (hoodmane) [inline packages/cryptography/meta.yaml:15]: Nice! How is this related to this PR? Could this change happen in a separate commit?

File: src/js/dynload.ts [hunk 2]
```diff
@@ -157,74 +200,9 @@ export async function loadDynlibsFromPackage(
     pkg.file_name.split("-")[0]
   }.libs`;
 
-  // This prevents from reading large libraries multiple times. We may read it
-  // once in calculateGlobalLibs and again in loadDynlib.
-  // Memoized function is intentionally instanced per call to
-  // loadDynlibsFromPackage since we won't need the data again after we're done.
-  const readFileMemoized: ReadFileType = memoize(Module.FS.readFile);
```
Comment (hoodmane) [inline src/js/dynload.ts:164]: Really great to get rid of this @ryanking13. Thanks for writing it and also for removing it =)

File: src/js/dynload.ts [hunk 3]
```diff
@@ -3,28 +3,53 @@
...
+    if (!Module.FS.isDir(mode)) {
+      continue;
+    }
+
+    subDirs.push(subdir);
+    subDirs.push(...getSubDirs(subdir));
+  }
+
+  return subDirs;
+}
```
Comment (hoodmane) [inline src/js/dynload.ts:41]: A minor optimization:
```suggestion
function getSubDirs(dir: string, subDirs: string[] = []): string[] {
  const dirs = Module.FS.readdir(dir);

  for (const d of dirs) {
    if (d === "." || d === "..") {
      continue;
    }

    const subdir = Module.PATH.join2(dir, d);
    const lookup = Module.FS.lookupPath(subdir);
    if (lookup.node === null) {
      continue;
    }

    const mode = lookup.node.mode;
    if (!Module.FS.isDir(mode)) {
      continue;
    }

    subDirs.push(subdir);
    getSubDirs(subdir, subDirs);
  }

  return subDirs;
}
```

Comment (hoodmane) [inline src/js/dynload.ts:41]: Actually this would be a great spot for a generator:
```suggestion
function* getSubDirs(dir: string): Iterator<string> {
  const dirs = Module.FS.readdir(dir);

  for (const d of dirs) {
    if (d === "." || d === "..") {
      continue;
    }

    const subdir = Module.PATH.join2(dir, d);
    const lookup = Module.FS.lookupPath(subdir);
    if (lookup.node === null) {
      continue;
    }

    const mode = lookup.node.mode;
    if (!Module.FS.isDir(mode)) {
      continue;
    }

    yield subdir;
    yield from getSubDirs(subdir);
  }
}
```

File: src/js/dynload.ts [hunk 4]
```diff
@@ -37,20 +62,35 @@ function createDynlibFS(
...
     for (const dir of _searchDirs) {
       const fullPath = Module.PATH.join2(dir, path);
+
       if (Module.FS.findObject(fullPath) !== null) {
         return fullPath;
       }
     }
+
+    // Step 2) try to find the library by searching child directories of the library directory
+    //         (This should not be necessary in most cases, but some libraries have dependencies in the child directories)
```
Comment (hoodmane) [inline src/js/dynload.ts:80]: This is odd. Does normal linux have logic like this? Or does it have some different behavior with `LD_LIBRARY_PATH` or something?

File: src/js/dynload.ts [hunk 5]
```diff
@@ -37,20 +62,35 @@ function createDynlibFS(
...
+    const childDirs = getSubDirs(dirname);
+    for (const childDir of childDirs) {
```
Comment (hoodmane) [inline src/js/dynload.ts:82]:
```suggestion
    for (const childDir of getSubDirs(dirname)) {
```

File: packages/netcdf4/meta.yaml
Comment (hoodmane) [inline packages/netcdf4/meta.yaml:1]: Presumably the changes to `netcdf4` can be split off and go in ahead of the rest of this PR?

File: packages/biopython/meta.yaml [hunk 6]
```diff
@@ -1,6 +1,7 @@
 package:
   name: biopython
   version: "1.83"
+  _disabled: true # setuptools v70.1.0 issue
```
Comment (hoodmane) [inline packages/biopython/meta.yaml:4]: Split off in #4877.

File: packages/openssl/meta.yaml
Comment (hoodmane) [inline packages/openssl/meta.yaml:1]: Again, would like to see these changes merged before or after the loader changes if possible. If not, would appreciate more detail on why the loader changes necessitated this. At least we should try not to update openssl at the same time as everything else.

File: packages/ssl/meta.yaml
Comment (hoodmane) [inline packages/ssl/meta.yaml:1]: Is there a functional change here or did you just tidy up? If this is an NFC tidy up would like to see it split off.

File: packages/cryptography/meta.yaml (hunk 1 above)
Comment (ryanking13) [reply packages/cryptography/meta.yaml:15]: I'll split off recipe changes into another PR.

**Comment by ryanking13** [comment]
>  The patch is so small for something that's caused us so much trouble!

Yes it is. Actually, this one (https://github.com/emscripten-core/emscripten/pull/18924) was a more critical fix (which is already included in Emscripten < 3.1.58), but it's also quite a small PR.

File: src/js/dynload.ts (hunk 3 above)
Comment (ryanking13) [reply src/js/dynload.ts:41]: Oh, right. I didn't think of using generators in JS. Thanks for the suggestion.

File: src/js/dynload.ts (hunk 4 above)
Comment (ryanking13) [reply src/js/dynload.ts:80]: Locating the path of the shared library is handled two ways in linux.

1. Using `LD_LIBRARY_PATH`
2. Using `rpath`

In emscripten, `LD_LIBRARY_PATH` is used [in _dlopen](https://github.com/emscripten-core/emscripten/blob/82b0d740924f589b320c624191497e1ab7857edc/system/lib/libc/dynlink.c#L518-L527), but we are not using it.

`rpath` is not supported in emscripten. I made a proposal to add it before (https://github.com/pyodide/pyodide/issues/3854), I guess I need to open a same proposal in emscripten too.

Comment (hoodmane) [reply src/js/dynload.ts:78]: Can we add a `// TODO: add rpath to Emscripten dsos and remove this logic`?

**Review by hoodmane — APPROVED** [review]
Thanks @ryanking13! Great work.

**Comment by hoodmane** [comment]
`duckdb/test_duckdb.py::duckdb_with_pandas[firefox]` times out a lot, we should xfail it.

**Comment by ryanking13** [comment]
> duckdb/test_duckdb.py::duckdb_with_pandas[firefox] times out a lot, we should xfail it.

Yes, agreed.

**Comment by ryanking13** [comment]
It seems like there is an additional error in the Scipy test suite. It's good to have a Scipy test so we can catch this.

```
2024-07-28T04:09:24.8596305Z Error: Dynamic linking error: cannot resolve symbol d_sign
2024-07-28T04:09:24.8597626Z     at stubs.<computed> (/home/runner/work/pyodide/pyodide/node_modules/pyodide/pyodide.asm.js:10:147647)
2024-07-28T04:09:24.8598748Z     at wasm://wasm/0003e6d6:wasm-function[97]:0x8d59
2024-07-28T04:09:24.8599535Z     at dynCall (/home/runner/work/pyodide/pyodide/node_modules/pyodide/pyodide.asm.js:10:145556)
2024-07-28T04:09:24.8600933Z     at /home/runner/work/pyodide/pyodide/node_modules/pyodide/pyodide.asm.js:10:145675
2024-07-28T04:09:24.8602645Z     at stubs.<computed> (/home/runner/work/pyodide/pyodide/node_modules/pyodide/pyodide.asm.js:10:147720)
2024-07-28T04:09:24.8603878Z     at wasm://wasm/0003e6d6:wasm-function[89]:0x2991
2024-07-28T04:09:24.8604701Z     at wasm://wasm/0003e6d6:wasm-function[107]:0xbfa0
2024-07-28T04:09:24.8605738Z     at wasm://wasm/0267c526:wasm-function[1143]:0x1c3691
2024-07-28T04:09:24.8606739Z     at wasm://wasm/0267c526:wasm-function[1144]:0x1c36c2
2024-07-28T04:09:24.8607550Z     at wasm://wasm/0267c526:wasm-function[3406]:0x29e6f6 {
2024-07-28T04:09:24.8608040Z   pyodide_fatal_error: true
2024-07-28T04:09:24.8608321Z }
```

**Comment by ryanking13** [comment]
The `d_sign` symbol comes from openblas, and 13 modules in scipy are importing it:

```
scipy/integrate/_dop.cpython-312-wasm32-emscripten.so:
       env      FUNC	d_sign
scipy/integrate/_lsoda.cpython-312-wasm32-emscripten.so:
       env      FUNC	d_sign
scipy/integrate/_odepack.cpython-312-wasm32-emscripten.so:
       env      FUNC	d_sign
scipy/integrate/_vode.cpython-312-wasm32-emscripten.so:
       env      FUNC	d_sign
scipy/interpolate/dfitpack.cpython-312-wasm32-emscripten.so:
       env      FUNC	d_sign
scipy/odr/__odrpack.cpython-312-wasm32-emscripten.so:
       env      FUNC	d_sign
scipy/optimize/_slsqp.cpython-312-wasm32-emscripten.so:
       env      FUNC	d_sign
scipy/special/_specfun.cpython-312-wasm32-emscripten.so:
       env      FUNC	d_sign
scipy/special/_ufuncs.cpython-312-wasm32-emscripten.so:
       env      FUNC	d_sign
scipy/special/cython_special.cpython-312-wasm32-emscripten.so:
       env      FUNC	d_sign
scipy/sparse/linalg/_propack/_dpropack.cpython-312-wasm32-emscripten.so:
       env      FUNC	d_sign
scipy/sparse/linalg/_propack/_zpropack.cpython-312-wasm32-emscripten.so:
       env      FUNC	d_sign
scipy/sparse/linalg/_eigen/arpack/_arpack.cpython-312-wasm32-emscripten.so:
       env      FUNC	d_sign
```

and among them, `_dop`, `_slsqp`, and `_specfun` modules are not linked to openblas correctly.

```
│   'scipy/integrate/_dop.cpython-312-wasm32-emscripten.so': [],
│   'scipy/integrate/_lsoda.cpython-312-wasm32-emscripten.so': ['libopenblas.so'],
│   'scipy/integrate/_odepack.cpython-312-wasm32-emscripten.so': ['libopenblas.so'],
│   'scipy/integrate/_vode.cpython-312-wasm32-emscripten.so': ['libopenblas.so'],
│   'scipy/interpolate/dfitpack.cpython-312-wasm32-emscripten.so': ['libopenblas.so'],
│   'scipy/odr/__odrpack.cpython-312-wasm32-emscripten.so': ['libopenblas.so'],
│   'scipy/optimize/_slsqp.cpython-312-wasm32-emscripten.so': [],
│   'scipy/special/_specfun.cpython-312-wasm32-emscripten.so': [],
│   'scipy/special/_ufuncs.cpython-312-wasm32-emscripten.so': ['libopenblas.so'],
│   'scipy/special/cython_special.cpython-312-wasm32-emscripten.so': ['libopenblas.so'],
│   'scipy/sparse/linalg/_propack/_dpropack.cpython-312-wasm32-emscripten.so': ['libopenblas.so'],
│   'scipy/sparse/linalg/_propack/_zpropack.cpython-312-wasm32-emscripten.so': ['libopenblas.so'],
│   'scipy/sparse/linalg/_eigen/arpack/_arpack.cpython-312-wasm32-emscripten.so': ['libopenblas.so'],
```

**Comment by ryanking13** [comment]
Okay, I think I figured out what is happening.

1. There are some symbols that come from `libf2c.a`, and scipy depends on it.
2. Currently, these symbols are provided to scipy modules indirectly, libf2c.a is statically linked to openblas.so and openblas.so is linked to scipy modules.
3. Some of scipy modules like `_dop.so` require symbols from `libf2c.a`, however they do not require symbols from openblas.so, so they are not linked to openblas.so.

The possible workarounds would be:

- Linking openblas.so to those modules or
- Linking libf2c.a to those modules

cc: @lesteve

**Comment by hoodmane** [comment]
Why not link `openblas` to these? Seems like the most reasonable approach to me.

**Comment by lesteve** [comment]
> It's good to have a Scipy test so we can catch this.

Great to hear that :heart: 

Other than this, no strong opinion/expertise on this so do whatever you think makes sense. I guess at the time I worked on OpenBLAS packaging I tried a few things and stuck with the first thing I managed to get working (probably https://github.com/pyodide/pyodide/pull/3331#discussion_r1160457087 is the most relevant comment related to this) ...

**Comment by ryanking13** [comment]
> Why not link openblas to these? Seems like the most reasonable approach to me.

Yes, [this patch](https://github.com/pyodide/pyodide/pull/4876/commits/78f257c3647b5da9929f5d9293f85a35dd8b033d), we may even link openblas.so to every C-extension modules, which would make our life easier... but it may impact the overall load time, so let me see if I can locate all fortran-based C-extension modules and link to them
